text/0156-columnar-encoding.md

```diff
@@ -0,0 +1,1417 @@
+more information on the implementation of this identifier.
+
+The `otlp_arrow_payloads` attribute is a list of `OtlpArrowPayload` messages. Each `OtlpArrowPayload` message represents
+a table of data encoded in a columnar format (e.g. metrics, logs, traces, and future OTLP entities). Several correlated
+IPC Arrow messages of different nature and with different schemas can be sent in the same OTLP batch identified by
+`batch_id` and thus be processed as one unit without complex logic in the collector or any other processing systems.
+More details on the `OtlpArrowPayload` columns in the section [Mapping OTEL entities to Arrow records](#mapping-otel-entities-to-arrow-records).
+
+More specifically, an `OtlpArrowPayload` protobuf message is defined as:
+
+```protobuf
+message OtlpArrowPayload {
+  // [mandatory] A unique id assigned to a sub-stream of the batch sharing the same schema, and dictionaries.
+  string sub_stream_id = 1;
+
+  // [mandatory] Type of the OTLP Arrow payload.
+  OtlpArrowPayloadType type = 2;
+
+  // [mandatory] Serialized Arrow Record Batch
+  // For a description of the Arrow IPC format see: https://arrow.apache.org/docs/format/Columnar.html#serialization-and-interprocess-communication-ipc
+  bytes record = 4;
+
+  // [mandatory]
+  CompressionMethod compression = 5;
+}
+
+enum OtlpArrowPayloadType {
+  METRICS = 0;
+  LOGS = 1;
+  SPANS = 2;
+}
+
+message EncodedData {
+  bytes ipc_message = 1;
+  bytes arrow_data = 2;
+}
+
+enum CompressionMethod {
+  NO_COMPRESSION = 0;
+  ZSTD = 1;
+}
+```
+
+The `sub_stream_id` attribute is a unique identifier for a sub-stream of `BatchArrowRecords` sharing the same schema inside
+the
+scope of the current stream. This id will be used receiver side to keep track of the schema and dictionaries for a
+specific
+type of Arrow Records. See the [Substream Id generation](#substream-id-generation) section for more information on the
+implementation
+of this identifier.
+
+The `OtlpArrowPayloadType` enum specifies the `type` of the payload.
+
+The `record` attribute is a binary representation of the Arrow RecordBatch.
+
+The `compression` attribute is a mandatory attribute that is used to define the compression algorithms for the different
+bytes buffer.
+
+ZSTD offers a training mode, which can be used to tune the algorithm for a selected type of data. The result of this
+training is a dictionary that can be used to compress the data. Using this [dictionary](http://facebook.github.io/zstd/#small-data)
+can dramatically improve the compression rate for small batches. This future development will build on both the gRPC
+stream approach used in this proposal and the ability to send a ZSTD dictionary over the OTLP Arrow stateful protocol,
+allowing us to train the ZSTD algorithm on the first batches and then update the configuration of the ZSTD
+encoder/decoder with an optimized dictionary.
```

Comment:
OK, it seems like the suggestion is to train at runtime and send the dictionary over the network. However, this is for future, so not used currently? Can we move this to a separate section? OTEPs have "Future Possibilities" exactly for this, so that it is clear what is the current design vs possible future extension.

Reply:
It is also not clear to me why we have the `compression` field now, when we don't implement the dictionary optimization. Why can't `compression` field be added in the future, when the new design with dictionary optimization is introduced?

Reply:
I moved this paragraphe into a new section named "Future Possibilities".

Reply:
Also the `compression` field no longer exists.